File: packages/health_connector_hk_ios/lib/src/mappers/health_record_mappers/exercise/exercise_route_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show ExerciseRoute, ExerciseRouteLocation, Length;
import 'package:health_connector_hk_ios/src/pigeon/health_connector_hk_ios_api.g.dart'
    show ExerciseRouteDto, ExerciseRouteLocationDto;
import 'package:meta/meta.dart' show internal;

/// Extension to convert [ExerciseRouteLocation] to [ExerciseRouteLocationDto].
@internal
extension ExerciseRouteLocationToDto on ExerciseRouteLocation {
  ExerciseRouteLocationDto toDto() {
    return ExerciseRouteLocationDto(
      time: time.millisecondsSinceEpoch,
      latitude: latitude,
      longitude: longitude,
      altitudeMeters: altitude?.inMeters,
      horizontalAccuracyMeters: horizontalAccuracy?.inMeters,
      verticalAccuracyMeters: verticalAccuracy?.inMeters,
    );
  }
}

/// Extension to convert [ExerciseRouteLocationDto] to [ExerciseRouteLocation].
@internal
extension ExerciseRouteLocationFromDto on ExerciseRouteLocationDto {
  ExerciseRouteLocation toDomain() {
    return ExerciseRouteLocation.internal(
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true),
      latitude: latitude,
      longitude: longitude,
      altitude: altitudeMeters != null ? Length.meters(altitudeMeters!) : null,
      horizontalAccuracy: horizontalAccuracyMeters != null
          ? Length.meters(horizontalAccuracyMeters!)
          : null,
      verticalAccuracy: verticalAccuracyMeters != null
          ? Length.meters(verticalAccuracyMeters!)
          : null,
    );
  }
}

/// Extension to convert [ExerciseRoute] to [ExerciseRouteDto].
@internal
extension ExerciseRouteToDto on ExerciseRoute {
  ExerciseRouteDto toDto() {
    return ExerciseRouteDto(
      locations: locations.map((l) => l.toDto()).toList(),
    );
  }
}

/// Extension to convert [ExerciseRouteDto] to [ExerciseRoute].
@internal
extension ExerciseRouteFromDto on ExerciseRouteDto {
  ExerciseRoute toDomain() {
    return ExerciseRoute.internal(
      locations.map((l) => l.toDomain()).toList(),
    );
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/ExerciseRouteMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.ExerciseRoute
import androidx.health.connect.client.units.Length
import com.phamtunglam.health_connector_hc_android.pigeon.ExerciseRouteDto
import com.phamtunglam.health_connector_hc_android.pigeon.ExerciseRouteLocationDto
import java.time.Instant

/**
 * Converts a Health Connect [ExerciseRoute] to [ExerciseRouteDto].
 */
internal fun ExerciseRoute.toDto(): ExerciseRouteDto = ExerciseRouteDto(
    locations = route.map { it.toDto() },
)

/**
 * Converts a Health Connect [ExerciseRoute.Location] to [ExerciseRouteLocationDto].
 */
internal fun ExerciseRoute.Location.toDto(): ExerciseRouteLocationDto = ExerciseRouteLocationDto(
    time = time.toEpochMilli(),
    latitude = latitude,
    longitude = longitude,
    altitudeMeters = altitude?.inMeters,
    horizontalAccuracyMeters = horizontalAccuracy?.inMeters,
    verticalAccuracyMeters = verticalAccuracy?.inMeters,
)

/**
 * Converts [ExerciseRouteDto] to Health Connect [ExerciseRoute].
 */
internal fun ExerciseRouteDto.toHealthConnect(): ExerciseRoute = ExerciseRoute(
    route = locations.map { it.toHealthConnect() },
)

/**
 * Converts [ExerciseRouteLocationDto] to Health Connect [ExerciseRoute.Location].
 */
internal fun ExerciseRouteLocationDto.toHealthConnect(): ExerciseRoute.Location =
    ExerciseRoute.Location(
        time = Instant.ofEpochMilli(time),
        latitude = latitude,
        longitude = longitude,
        altitude = altitudeMeters?.let { Length.meters(it) },
        horizontalAccuracy = horizontalAccuracyMeters?.let {
            Length.meters(it)
        },
        verticalAccuracy = verticalAccuracyMeters?.let {
            Length.meters(it)
        },
    )
